src/System.Runtime.Analyzers/Core/DoNotRaiseReservedExceptionTypes.cs

```diff
@@ -35,14 +70,74 @@ public abstract class DoNotRaiseReservedExceptionTypesAnalyzer : DiagnosticAnaly
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             helpLinkUri: "https://msdn.microsoft.com/en-us/library/ms182338.aspx",
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(TooGenericRule, ReservedRule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(TooGenericRule, ReservedRule); 
+
+        public abstract TLanguageKindEnum ObjectCreationExpressionKind { get; }
+
+        public abstract SyntaxNode GetTypeSyntaxNode(TObjectCreationExpressionSyntax node);
 
         public override void Initialize(AnalysisContext analysisContext)
-        { 
-            
+        {
+                
+            analysisContext.RegisterCompilationStartAction(
+            compilationStartContext =>
+            {
+                HashSet<INamedTypeSymbol> tooGenericExceptionSymbols = CreateSymbolSet(compilationStartContext.Compilation, s_tooGenericExceptions);
+                HashSet<INamedTypeSymbol> reservedExceptionSymbols = CreateSymbolSet(compilationStartContext.Compilation, s_reservedExceptions); ;             
+
+                if (tooGenericExceptionSymbols.Count == 0 && reservedExceptionSymbols.Count == 0)
+                {
+                    return;
+                }
+
+                compilationStartContext.RegisterSyntaxNodeAction(
+                syntaxNodeContext =>
+                {
+                    Analyze(syntaxNodeContext, tooGenericExceptionSymbols, reservedExceptionSymbols);
+                },
+                ObjectCreationExpressionKind);
+            });    
+        }
+
+        private static HashSet<INamedTypeSymbol> CreateSymbolSet(Compilation compilation, IEnumerable<string> exceptionNames)
+        {
+            HashSet< INamedTypeSymbol> set = null;
+            foreach (var exp in exceptionNames)
+            {
+                var symbol = compilation.GetTypeByMetadataName(exp);
+                if (symbol == null)
+                {
+                    continue;
+                }
+                if (set == null)
+                {
+                    set = new HashSet<INamedTypeSymbol>();
+                }
+                set.Add(symbol);
+            }
+            return set;
+        }
+
+        private void Analyze(SyntaxNodeAnalysisContext context, HashSet<INamedTypeSymbol> tooGenericExceptionSymbols, HashSet<INamedTypeSymbol> reservedExceptionSymbols)
+        {
+            var objectCreationNode = (TObjectCreationExpressionSyntax) context.Node;
+            var targetType = GetTypeSyntaxNode(objectCreationNode);
+
+            // TODO: can this be null in error condition?
```

Comment:
However, it can cause GetSymbol().Symbol below to return an error type symbol - you probably want an "as INamedTypeSymbol" followed by a null check.